homeassistant/components/zha/entity.py

```diff
@@ -66,20 +66,34 @@ def __init__(self, entity_data: EntityData, *args, **kwargs) -> None:
 
     @cached_property
     def name(self) -> str | UndefinedType | None:
-        """Return the name of the entity."""
+        """Return the name of the entity.
+
+        Built-in quirks have translations in HA, so those are used.
+        Custom quirks with new translation keys won't have translations.
+        For them, the fallback name should be used instead.
+        If a device class is set but no translation key,
+        the device class name is used.
+        """
         meta = self.entity_data.entity.info_object
         if meta.primary:
             self._attr_name = None
             return super().name
 
-        original_name = super().name
-
-        if original_name not in (UNDEFINED, None) or meta.fallback_name is None:
-            return original_name
-
-        # This is to allow local development and to register niche devices, since
-        # their translation_key will probably never be added to `zha/strings.json`.
-        self._attr_name = meta.fallback_name
+        # if we have a translation key or no (device class) name,
+        # use translation or fallback name
+        if meta.translation_key is not None or super().name in (UNDEFINED, None):
+            # if we have a translation for this key, use it
+            if (
+                (translation_key := self._name_translation_key) is not None
+                and translation_key in self.platform_data.platform_translations
+            ):
+                return super().name
+            # a translation won't be available for custom quirks, so use fallback_name
+            if meta.fallback_name is not None:
+                self._attr_name = meta.fallback_name
+                return super().name
```

Comment:
note: I am aware this line could be removed, as we'd just hit the line below, doing the same, but I think keeping this makes it easier to see what's going on, especially when also looking at the comments.

There should be no real downside to keeping this.